Add the ability to verify/unverify questions

daltojohnso opened the pull request:
This PR adds the concept of "verification" to saved questions.

Admins will now see a "moderation" section in the saved question details sidebar (which is added in this PR https://github.com/metabase/metabase/pull/16803):

![Screen Shot 2021-07-19 at 4 16 04 PM](https://user-images.githubusercontent.com/13057258/126239221-3b24f3d4-0b0b-4caa-85ed-4d9eb0c7a7f3.png)

---------

After clicking the icon button to verify the saved question, a badge will appear next to the saved question's name and a banner will appear in the sidebar:

![Screen Shot 2021-07-19 at 4 01 55 PM](https://user-images.githubusercontent.com/13057258/126238986-14352df2-d7d6-4936-bb11-9f981bd20249.png)

---------

Verification is removable by clicking a close icon that appears in the banner when the user hovers over the banner with their mouse or focuses the button:

![Screen Shot 2021-07-19 at 4 02 18 PM](https://user-images.githubusercontent.com/13057258/126239083-abebbb32-e105-4016-abb9-8a9d515ca31a.png)

---------

Normal/non-admin users see nothing but the badge and the banner:
![Screen Shot 2021-07-19 at 4 02 46 PM](https://user-images.githubusercontent.com/13057258/126239101-316736c3-05fd-450f-8fd0-01ff32cd6df4.png)

---------

Here's a screen recording of what it looks like from the moderator's perspective when verifying a saved question:

https://user-images.githubusercontent.com/13057258/126017192-949c57cd-a62b-4d6f-b3f3-08dd102ca05d.mov

daltojohnso committed:
rmv old bucm icons and remove verified fill color

daltojohnso committed:
add moderation action section to sidebar

daltojohnso committed:
add moderation review icon to the saved question header button

daltojohnso committed:
hide moderation section when is not a moderator

daltojohnso committed:
add UI for ModerationReviewBanner

dpsutton committed:
Backend for moderation-review

- create table moderation_review. Same as before but also has a
  "most_recent" boolean flag for the most recent moderation for easy
  lookup
- POST /moderation-review/ . Status can be "verified" or nil
- must be an admin to post
- No PUT or edit route yet. Not sure if this is even
  necessary. _MAYBE_ to edit the text, but certainly not for the
  status, ids, etc. If there's to be history, let's build some history
- Ensure we never have more than 10 reviews. Adding a new review will
  delete the older ones, mark all old ones as not `most_recent`, and
  add the newest one as `most_recent true`
- Ensure the card actually exists before creating the mod review
- Since admin only at this time, don't need to check moderate
  permission or view permission
- When hydrating ensure reviews are ordered by id desc. Should mimic
  the created_at desc

daltojohnso committed:
fix moderation review banner tooltip offset

daltojohnso committed:
disable verification button when already verified

daltojohnso committed:
rmv iconOnly prop because it seems to do nothing

daltojohnso committed:
update getLatestModerationReview to rely on most_recent boolean

dpsutton committed:
Return 400 on invalid status to post /moderation-review

the schema was using keywords on the left hand side rather than the
symbols. Required a change to the docstring generator, when it made a
docstring for enums, it would call (sort (:vs enum)) and need to
string em.

dpsutton committed:
Add ModerationReview model to models.clj and copy infra

dpsutton committed:
hydrate moderation reviews on cards

daltojohnso committed:
clean up + wire up to BE + ensure mod buttons don't show for normal users

daltojohnso committed:
rmv unused moderation redux logic from QuestionDetailsSidebarPanel

daltojohnso committed:
finish writing unit tests for FE

daltojohnso committed:
ensure getIconForReview returns an object

daltojohnso committed:
enable/disable verify button tooltip when unverified/verified

daltojohnso committed:
add e2e tests

daltojohnso committed:
fix tests

mazameli committed:
styling tweaks

mazameli commented:
@daltojohnso can we have this disabled button not appear here when the question is verified?

![image](https://user-images.githubusercontent.com/2223916/126577208-22b8cc48-dd2f-4e1b-8a1c-2aa2c6fb7c3b.png)

Also just an FYI that I pushed a couple style tweaks: changed the label text treatment to match the "History" heading, and fixed the sizing on the Verify button to match the other buttons:

![image](https://user-images.githubusercontent.com/2223916/126577155-6ce5a4b4-6faa-42a9-bf69-e8a596013ef9.png)

mazameli committed:
more styling on moderationReviewBanner

mazameli committed:
add function for abbreviated timestamp

mazameli committed:
increase fontsize of timestamp back to 12

mazameli commented on frontend/src/metabase/lib/time.js:
```diff
@@ -119,3 +119,29 @@ export function formatFrame(frame) {
+  return moment(timestamp).fromNow();
+}
+
+export function getRelativeTimeAbbreviated(timestamp) {
```
@daltojohnso I implemented this to get the timestamps to look like `19 m` based on a thread I saw in on a Moment.js GH issue, but let me know if this is hacky or something.

mazameli commented:
Looking pretty close to the designs now (browser on left, designs on right):

![image](https://user-images.githubusercontent.com/2223916/126682250-4b470c1a-6314-4efb-a9cc-e4eb5faa4db2.png)

daltojohnso committed:
fix tooltip offset

daltojohnso committed:
ensure custom locale is separate from 'en' and not used for other languages

daltojohnso commented on frontend/src/metabase/lib/time.js:
```diff
@@ -123,3 +155,15 @@ export function formatFrame(frame) {
 }
+
+export function getRelativeTimeAbbreviated(timestamp) {
+  const locale = moment().locale();
```
This code is really annoying to write unit tests for because of the way `moment` handles `locale` globally. If anyone would like, I can do something hacky like `export const _moment_from_module_ = moment` so that the unit test has access to the same version of `moment` as the module. Otherwise, we can wait for the Jest update to go through and mocking will become easier with `jest.requireActual`

mazameli commented:
@metabase/frontend-developers, I believe @daltojohnso had intended to rebase this on master on Friday, and then get code review on it. He's out until Thursday, so could I get someone to A) see if this has indeed been rebased, and B) give the code a review?

dpsutton committed:
Deletion moderation reviews when deleting cards

i had actually thought this was a much larger problem. But it turns
out we almost never delete cards (thanks comment!). And so we won't
really generate a lot of garbage.

I was worried that since we aren't using actual foreign keys but just
`moderated_item_type "card"` and `moderated_item_id 2` we would have
deleted cards with these moderation reviews but that is not the case
as the cards aren't deleted.

alxnddr committed:
hide verify disabled button when a question is verified

alxnddr reviewed (approved):
Checked the FE code, it looks good. Added one suggestion and pushed a commit with hiding moderation button instead of disabling it as @mazameli suggested

mazameli commented on enterprise/frontend/src/metabase-enterprise/moderation/components/ModerationActions/ModerationActions.unit.spec.js:
```diff
@@ -0,0 +1,26 @@
+  describe("when the user is not a moderator", () => {
+    it("should not render", () => {
+      const { getByTestId } = render(<ModerationActions isModerator={false} />);
+      expect(() => getByTestId("moderation-verify-action")).toThrow();
```
Thanks, @alxnddr!

daltojohnso replied:
Good idea -- updated that test

daltojohnso committed:
update test to use queryByTestId

dpsutton committed:
Hydrate moderation reviews on cards on ordered cards

dpsutton committed:
Handle mysql's lack of offset functionality

mysql cannot handle just a `offset` clause, it also needs a limit

clause grammar from
https://dev.mysql.com/doc/refman/8.0/en/select.html:

[LIMIT {[offset,] row_count | row_count OFFSET offset}]

select id, name from metabase_field offset 5;         -- errors
select id, name from metabase_field limit 2 offset 5; -- works

Since our numbers are so small here there is no worry and just do the
offset in memory rather than jump through hoops for different dbs.

mazameli commented:
One thing I just saw: we shouldn't display this block to non-admins:

![image](https://user-images.githubusercontent.com/2223916/127690095-e999d0bb-1249-4020-9567-2a85feca31c4.png)

dpsutton committed:
Batch hydrate moderation reviews

daltojohnso committed:
Don't let /api/user/:userId failure conceal moderation banner

daltojohnso committed:
fix moderation cy tests

dpsutton committed:
work around possible bug in toucan hydration

dashboards hydrate ordered cards
(hydrate [:ordered_cards [:card :moderation_reviews] :series])

Ordered_cards are dashboard_cards which have an optional card_id. But
toucan hydration doesn't filter out the nils as they go down. It seems
toucan returns a nil card, and then when hydrating the
moderation_review, passes the collection of all "cards" including the
nil ones into the hydration function for moderation_reviews

This feels like a bug to me

dpsutton committed:
Cleanup moderation warnings

dpsutton committed:
Docstring in moderation review

dpsutton committed:
include hoisted moderated_status on cards in collections api

dpsutton committed:
Expect unverified in test :wrench:

jeff303 commented on test/metabase/api/moderation_review_test.clj:
```diff
@@ -0,0 +1,86 @@
+                                         {:text                "card doesn't exist"
+                                          :status              "verified"
+                                          :moderated_item_id   Integer/MAX_VALUE
+                                          :moderated_item_type "card"})))))))))
```
Is there a test that unverify works (i.e. the last part showed in Dalton's screen recording in the comments)?

dpsutton replied:
i don't think there is. i'll add one. thanks @jeff303

daltojohnso replied:
we'll move the writing of this test to the next BUCM branch so that we can unblock the next batch of BUCM work
